### PR TITLE
Omlt integration

### PR DESCRIPTION
**Pull request by carldlaird**
## NOTE:
This PR skips all of the tests in test_commands.py so that the other tests could run. This will need to be fixed ASAP.

## Fixes
- Changed the way options are passed to derived SurrogateObjects from SurrogateBlock so we can track unused options. This API is internal and should not impact users.
- Added checks to make sure that the keys in the input_bounds match the input_labels
- Added the KerasSurrogate object that provides a link between IDAES and OMLT for Keras Sequential models.
- Exposed the split_dataframe function (moved from private to public) to support some generic splitting of data frames
- Added an offset/factor based scaling object (OffsetScaler) to support easy scaling and unscaling of data in a dataframe. This is also used in the KerasSurrogate to provide the scaling factors to the OMLT scaling objects.
- Added the script and data that can be used to train the Keras models used in the tests (create_keras_models) in tests/data
- Added some pre-trained Keras models that can be used in testing
- Added tensorflow and OMLT to the dependencies.

## Summary/Motivation:
This PR provides the ability to import Keras Sequential models into IDAES using OMLT

## Changes proposed in this PR:
-
-

### Legal Acknowledgement

By contributing to this software project, I agree to the following terms and conditions for my contribution:

1. I agree my contributions are submitted under the license terms described in the LICENSE.txt file at the top level of this directory.
2. I represent I am authorized to make the contributions and grant the license. If my employer has rights to intellectual property that includes these contributions, I represent that I have received permission to make contributions and grant the required license on behalf of that employer.

### CODE REVIEW

File: idaes/surrogate/keras_surrogate.py [hunk 1]
```diff
@@ -0,0 +1,201 @@
...
+    def save_to_file(cls, keras_folder_name):
+        """
+        Save the surrogate object to disk by providing the name of the
+        folder to contain the keras model and additional IDAES metadata
+
+        Args:
+           folder_name: str
+              The name of the folder to contain the Keras model and additional
+              IDAES metadata
+        """
```
Comment (jsiirola) [inline idaes/surrogate/keras_surrogate.py:149]: This doesn't need to change for this PR, but I think that long term this method should:
- create a temporary directory, 
- pass that dir to `_keras_model.save()`, 
- add the `idaes_info.json` to the dir,
- zip up the dir to the file name passed to `save_to_file`
- delete the temporary directory

File: idaes/surrogate/keras_surrogate.py [hunk 2]
```diff
@@ -0,0 +1,201 @@
...
+            json.dump(info, fd)
+
+    @classmethod
+    def load_from_file(cls, keras_folder_name):
+        """
+        Load the surrogate object from disk by providing the name of the
+        folder holding the keras model
+
+        Args:
+           folder_name: str
```
Comment (jsiirola) [inline idaes/surrogate/keras_surrogate.py:174]: Similar to my comment above, I think this should take an archive file (ZIP?) as the argument and not a directory.

File: idaes/surrogate/sampling/scaling.py [hunk 3]
```diff
@@ -0,0 +1,142 @@
...
+# Lawrence Berkeley National Laboratory,  National Technology & Engineering
+# Solutions of Sandia, LLC, Carnegie Mellon University, West Virginia University
+# Research Corporation, et al.  All rights reserved.
+#
+# Please see the files COPYRIGHT.md and LICENSE.md for full copyright and
+# license information.
+#################################################################################
+import pandas as pd
+
+class StandardScaler(object):
```
Comment (jsiirola) [inline idaes/surrogate/sampling/scaling.py:15]: Would a better name be `StandardDeviationScalar` or `MeanStdDevScalar`?  Or do you intend this to be the "standard" scalar?

**Comment by andrewlee94** [comment]
@carldlaird It also looks like we need to add some `skipif` statements to the Keras tests as well as get Keras installed as part of the CI environment for these.

File: idaes/surrogate/keras_surrogate.py [hunk 4]
```diff
@@ -0,0 +1,201 @@
...
+"""
+import numpy as np
+import json
+import os.path
+import keras
+from idaes.surrogate.base.surrogate_base import SurrogateBase
+from idaes.surrogate.sampling.scaling import StandardScaler
+from omlt import OmltBlock, OffsetScaling
+from omlt.neuralnet import FullSpaceContinuousFormulation, ReducedSpaceContinuousFormulation, \
+    ReLUBigMFormulation, load_keras_sequential
```
[the remark this answers is not in the document]
Comment (carldlaird) [reply idaes/surrogate/keras_surrogate.py]: I made this change

File: idaes/surrogate/keras_surrogate.py [hunk 5]
```diff
@@ -0,0 +1,201 @@
...
+        net = load_keras_sequential(self._keras_model, omlt_scaling, self.input_bounds())
+
+        if formulation == 'full-space':
+            formulation_object = FullSpaceContinuousFormulation(net)
+        elif formulation == 'reduced-space':
+            formulation_object = ReducedSpaceContinuousFormulation(net)
+        elif formulation == 'relu-bigm':
+            formulation_object = ReLUBigMFormulation(net)
+        elif formulation == 'relu-complementarity':
+            formulation_object = ReLUComplementarityFormulation(net)
```
[the remark this answers is not in the document]
Comment (carldlaird) [reply idaes/surrogate/keras_surrogate.py:139]: I made this change.

File: idaes/surrogate/keras_surrogate.py (hunk 1 above)
Comment (carldlaird) [reply idaes/surrogate/keras_surrogate.py:181]: I agree. I also found a new way to store and read keras models (used for testing now). We might be able to build off of that.

File: idaes/surrogate/sampling/scaling.py (hunk 3 above)
Comment (carldlaird) [reply idaes/surrogate/sampling/scaling.py]: I changed this to OffsetScaler

**Comment by carldlaird** [comment]
> @carldlaird It also looks like we need to add some `skipif` statements to the Keras tests as well as get Keras installed as part of the CI environment for these.

Yep. I think I got both of these.

File: idaes/surrogate/keras_surrogate.py [hunk 6]
```diff
@@ -0,0 +1,201 @@
...
+
+        block.neural_network.build_formulation(formulation_object,
+                                               input_vars=block._input_vars_as_list(),
+                                               output_vars=block._output_vars_as_list())
```
[the remark this answers is not in the document]
Comment (carldlaird) [reply idaes/surrogate/keras_surrogate.py]: This has been fixed.

File: idaes/surrogate/keras_surrogate.py (hunk 2 above)
Comment (carldlaird) [reply idaes/surrogate/keras_surrogate.py:206]: I agree we should look at how we serialize all of the surrogate objects in a future PR.

**Review by jsiirola — APPROVED** [review]
Several questions / comments, but nothing that would prevent merging.

File: idaes/commands/tests/test_commands.py [hunk 7]
```diff
@@ -26,6 +26,7 @@
 # third-party
 from click.testing import CliRunner
 import pytest
+pytestmark = pytest.mark.skip
```
[the remark this answers is not in the document]
Comment (ksbeattie) [reply idaes/commands/tests/test_commands.py:29]: This was on my recommendation.  It's because of #493 which I hope to have addressed in #629

**Review by MAZamarripa — APPROVED** [review]
Code works fine
I tested build KerasSurrogate, keras_surrogate.evaluate_surrogate, and build_model
I used our MEA example to do the test. 
Improvements:
- input_scaler was a bit difficult, specially because normalize data like this:
xdata[i, j] = (xdata[i, j] - xmin[j])/(xmax[j] - xmin[j])
- it worked for sequential models, I think there was an error when I tried a different model
- I think we had some issues with the CustomLayer and network structure (we will provide code for additional testing).
- I had issues using save and load_keras_json_hd5

File: idaes/surrogate/keras_surrogate.py [hunk 8]
```diff
@@ -0,0 +1,242 @@
...
+        Args:
+           block: Pyomo Block component
+              The block to be populated with variables and/or constraints.
+           additional_options: dict or None
+              If not None, then should be a dict with the following keys:
+                 'formulation': KerasSurrogate.Formulation
+                    The formulation to use with OMLT. Possible values are FULL_SPACE,
+                    REDUCED_SPACE, RELU_BIGM, or RELU_COMPLEMENTARITY (default is FULL_SPACE)
+        """
+        formulation = additional_options.pop('formulation', KerasSurrogate.Formulation.FULL_SPACE)
```
[the remark this answers is not in the document]
Comment (carldlaird) [reply idaes/surrogate/keras_surrogate.py:109]: Interesting. I will make this change in a future PR. I like it.

File: idaes/surrogate/keras_surrogate.py [hunk 9]
```diff
@@ -0,0 +1,242 @@
...
+            x = self._input_scaler.scale(x)
+        y = self._keras_model.predict(x.to_numpy())
+
+        # y is a numpy array, make it a dataframe
+        y = pd.DataFrame(data=y, columns=self.output_labels(), index=inputs.index)
+        if self._output_scaler is not None:
+            y = self._output_scaler.unscale(y)
+        return y
+
+    def save_to_folder(self, keras_folder_name):
```
[the remark this answers is not in the document]
Comment (carldlaird) [reply idaes/surrogate/keras_surrogate.py:172]: This is by design until we finalize the way to save this to a single file. I was using the same "save_to_file" except that it isn't a file, and I thought that would be more confusing.

**Comment by carldlaird** [comment]
> Code works fine I tested build KerasSurrogate, keras_surrogate.evaluate_surrogate, and build_model I used our MEA example to do the test. Improvements:
> 
> * input_scaler was a bit difficult, specially because normalize data like this:
>   xdata[i, j] = (xdata[i, j] - xmin[j])/(xmax[j] - xmin[j])
> * it worked for sequential models, I think there was an error when I tried a different model
> * I think we had some issues with the CustomLayer and network structure (we will provide code for additional testing).
> * I had issues using save and load_keras_json_hd5

Thanks, @MAZamarripa. Let's discuss these for a future PR.

File: idaes/surrogate/base/surrogate_base.py [hunk 10]
```diff
@@ -243,21 +258,18 @@ def input_bounds(self):
             return dict(self._input_bounds)
         return None
     
-    def populate_block(self, block, **kwargs):
+    def populate_block(self, block, additional_options=None):
```
[the remark this answers is not in the document]
Comment (carldlaird) [reply idaes/surrogate/base/surrogate_base.py:261]: However, the approach I implemented causes problems with documentation. We should have a design discussion about how to best deal with this.
